server/api/db/sqldb/db.py

```diff
@@ -1121,47 +1121,57 @@ def _mark_best_iteration_artifact(
         if link_key:
             key = link_key
 
-        # We perform two consecutive SELECT queries and modify the two artifact records that are returned.
-        # Without no_autoflush, SQLAlchemy offloads the transaction data to the DB during the second SELECT query,
-        # which can result in a deadlock due to assumed conflicts between the transactions.
-        # Using no_autoflush ensures that all modifications don't arrive to the DB until the transaction is committed.
-        with session.no_autoflush:
-            # get the best iteration artifact record
-            query = self._query(session, ArtifactV2).filter(
-                ArtifactV2.project == project,
-                ArtifactV2.key == key,
-                ArtifactV2.iteration == link_iteration,
-            )
-            if link_tree:
-                query = query.filter(ArtifactV2.producer_id == link_tree)
-            if uid:
-                query = query.filter(ArtifactV2.uid == uid)
-
-            best_iteration_artifact_record = query.one_or_none()
-            if not best_iteration_artifact_record:
-                raise mlrun.errors.MLRunNotFoundError(
-                    f"Best iteration artifact not found - {project}/{key}:{link_iteration}",
-                )
+        # Lock the artifacts with the same project and key (and producer_id when available) to avoid unexpected
+        # deadlocks and conform to our lock-once-when-starting logic - ML-6869
+        lock_query = self._query(
+            session,
+            ArtifactV2,
+            project=project,
+            key=key,
+        ).with_entities(ArtifactV2.id)
+        if link_tree:
+            lock_query = lock_query.filter(ArtifactV2.producer_id == link_tree)
+
+        lock_query.order_by(
+            ArtifactV2.id.asc()
+        ).populate_existing().with_for_update().all()
```

Comment:
Do we really need to order the query? Why?

Reply:
tbh I think we don't but we do it in tag artifacts where we lock them there. I think it's a sanity in case the DB locks the artifacts one by one it makes the locking deadlock proof if its not atomic